pkg/types/currency_pair.go

```diff
@@ -18,19 +20,135 @@ func NewCurrencyPair(base, quote string) CurrencyPair {
 	}
 }
 
+// IsLegacyAssetString returns true if the asset string is of the following format:
+// - contains no instances of fieldSeparator.
+func IsLegacyAssetString(asset string) bool {
+	return !strings.Contains(asset, fieldSeparator)
+}
+
+// coreValidate performs checks that are universal across any ticker format style, namely:
+// - check that base and quote are not empty.
+// - check that the length of the base and quote fields do not exceed MaxCPFieldLength.
+func (cp *CurrencyPair) coreValidate() error {
+	if cp.Base == "" {
+		return fmt.Errorf("base asset cannot be empty")
+	}
+
+	if cp.Quote == "" {
+		return fmt.Errorf("quote assetcannot be empty")
+	}
+
+	if len(cp.Base) > MaxCPFieldLength {
+		return fmt.Errorf("base asset exceeds max length of %d", MaxCPFieldLength)
+	}
+
+	if len(cp.Quote) > MaxCPFieldLength {
+		return fmt.Errorf("quote asset exceeds max length of %d", MaxCPFieldLength)
+	}
+
+	return nil
+}
+
 // ValidateBasic checks that the Base / Quote strings in the CurrencyPair are formatted correctly, i.e.
-// Base + Quote are non-empty, and are in upper-case.
+// For base and quote asset:
+// check if the asset is formatted in the legacy validation form:
+//   - if so, check that fields are not empty and all upper case
+//   - else, check that the format is in the following form: tokenName,tokenAddress,chainID
 func (cp *CurrencyPair) ValidateBasic() error {
+	if err := cp.coreValidate(); err != nil {
+		return err
+	}
+
+	if IsLegacyAssetString(cp.Base) {
+		err := ValidateLegacyAssetString(cp.Base)
+		if err != nil {
+			return fmt.Errorf("base asset %q is invalid: %w", cp.Base, err)
+		}
+	} else {
+		err := ValidateDefiAssetString(cp.Base)
+		if err != nil {
+			return fmt.Errorf("base defi asset %q is invalid: %w", cp.Base, err)
+		}
+	}
+
+	// check quote asset
+	if IsLegacyAssetString(cp.Quote) {
+		err := ValidateLegacyAssetString(cp.Quote)
+		if err != nil {
+			return fmt.Errorf("quote asset %q is invalid: %w", cp.Quote, err)
+		}
+	} else {
+		err := ValidateDefiAssetString(cp.Quote)
+		if err != nil {
+			return fmt.Errorf("quote defi asset %q is invalid: %w", cp.Quote, err)
+		}
+	}
+
+	return nil
+}
+
+// ValidateLegacyAssetString checks if the asset string is formatted correctly, i.e.
+// - asset string is fully uppercase
+// - asset string does not contain the `fieldSeparator`
+//
+// NOTE: this function assumes that coreValidate() has already been run.
+func ValidateLegacyAssetString(asset string) error {
```

Comment:
if someone were to use this function outside of this package, then `coreValidate` won't be visible to them. what do they need to run to trigger a core validate?